Allow airflow standard images to run in openshift utilising the official helm chart #18136

nwalens opened the pull request:
<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of existing issue, reference it using one of the following:

closes: #18136 

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->
This pull request adds the parameter rbac.createSCCRoleBinding.
When enabled, a new RoleBinding object will be created targeting the various serviceAccounts utilised by deployments and pods.
The roleBinding will target the ``system:openshift:scc:anyuid`` cluster role which allows for pods to start with any arbitrary uid.
The ideal solution would also add the possibility of removing the security contexts altogether, however this would not be possible due to a number of pod templates setting different uids.
I will investigate the possibility to add the option of removing the security contexts so that openshift can then set its own uid as intended.
As it is, the pods will start and work as expected, utilising the predefined uids set in values file or the default image uid as set during build.

The change is not intrusive and should work in the current workflow as is. The option is also set to false by default in order to not impact any existing setups. 

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

nwalens committed:
Add support for openshift security context constraints.

- Create a rolebinding for all required service accounts so they can use the anyuid SCC in openshift.

Signed-off-by: Ney walens De Mesquita <walens@gmail.com>

nwalens committed:
add extra option for security context constraints

- Support openshift securityContextConstraints in order to allow pods to run with anyuid scc.
- The option is disabled by default since it is specific to openshift 4 setups

Signed-off-by: Ney Walens De Mesquita <walens@gmail.com>

nwalens committed:
Cleanup on scc rolebing

- SCCs will always be cluster roles

Signed-off-by: Ney Walens De Mesquita <walens@gmail.com>

nwalens committed:
rename helm option and add further checks

- simplify the parameter name
- add checks for service accounts so we do not bind the role to unexisting accounts

Signed-off-by: Ney Walens De Mesquita <walens@gmail.com>

nwalens committed:
include documentation for the added parameter

Signed-off-by: Ney Walens De Mesquita <walens@gmail.com>

nwalens committed:
update changelog

- Signed-off-by: Ney Walens De Mesquita <walens@gmail.com>

jedcunningham commented on chart/CHANGELOG.txt:
```diff
@@ -3,6 +3,7 @@ Airflow Helm Chart 1.1.0, 2021-07-26
 
 New Features
 """"""""""""
+- Chart: Add support for running pods with anyuid ``securityContextConstraint`` (SCC) (#18136)
```
```suggestion
```

1.1.0 is already released, this will land in 1.2.0 and we will build the changelog then. Thanks for trying to be proactive though!

jedcunningham commented on chart/templates/rbac/security-context-constraint-rolebinding.yaml:
```diff
@@ -0,0 +1,89 @@
+    name: {{ include "scheduler.serviceAccountName" . }}
+    namespace: {{ .Release.Namespace }}
+{{- end }}
+{{- if .Values.statsd.enabled }}
```
```suggestion
{{- if and .Values.statsd.enabled .Values.statsd.serviceAccount.create }}
```

jedcunningham commented on chart/templates/rbac/security-context-constraint-rolebinding.yaml:
```diff
@@ -0,0 +1,89 @@
+    name: {{ include "worker.serviceAccountName" . }}
+    namespace: {{ .Release.Namespace }}
+{{- end }}
+{{- if has .Values.executor $schedulerLaunchExecutors }}
```
```suggestion
{{- if and (has .Values.executor $schedulerLaunchExecutors) .Values.scheduler.serviceAccount.create }}
```

jedcunningham commented on chart/templates/rbac/security-context-constraint-rolebinding.yaml:
```diff
@@ -0,0 +1,89 @@
+  - kind: ServiceAccount
+    name: {{ include "webserver.serviceAccountName" . }}
+    namespace: {{ .Release.Namespace }}
+{{- if has .Values.executor $workerLaunchExecutors }}
```
```suggestion
{{- if and (has .Values.executor $workerLaunchExecutors) .Values.workers.serviceAccount.create }}
```

jedcunningham commented on chart/templates/rbac/security-context-constraint-rolebinding.yaml:
```diff
@@ -0,0 +1,89 @@
+  kind: ClusterRole
+  name: system:openshift:scc:anyuid
+subjects:
+  - kind: ServiceAccount
+    name: {{ include "webserver.serviceAccountName" . }}
+    namespace: {{ .Release.Namespace }}
```
```suggestion
{{- if .Values.webserver.serviceAccount.create }}
  - kind: ServiceAccount
    name: {{ include "webserver.serviceAccountName" . }}
    namespace: {{ .Release.Namespace }}
{{- end }}    
```

jedcunningham commented on chart/templates/rbac/security-context-constraint-rolebinding.yaml:
```diff
@@ -0,0 +1,89 @@
+    name: {{ include "statsd.serviceAccountName" . }}
+    namespace: {{ .Release.Namespace }}
+{{- end }}
+  - kind: ServiceAccount
+    name: {{ include "flower.serviceAccountName" . }}
+    namespace: {{ .Release.Namespace }}
```
```suggestion
{{- if and .Values.flower.enabled (or (eq .Values.executor "CeleryExecutor") (eq .Values.executor "CeleryKubernetesExecutor")) .Values.flower.serviceAccount.create }}
  - kind: ServiceAccount
    name: {{ include "flower.serviceAccountName" . }}
    namespace: {{ .Release.Namespace }}
{{- end }}
```

jedcunningham commented on chart/templates/rbac/security-context-constraint-rolebinding.yaml:
```diff
@@ -0,0 +1,89 @@
+    name: {{ include "createUserJob.serviceAccountName" . }}
+    namespace: {{ .Release.Namespace }}
+{{- end }}
+{{- if .Values.cleanup.enabled }}
```
```suggestion
{{- if and .Values.cleanup.enabled .Values.cleanup.serviceAccount.create }}
```

jedcunningham commented on chart/values.yaml:
```diff
@@ -1251,7 +1252,7 @@ dags:
     size: 1Gi
     # If using a custom storageClass, pass name here
-    storageClassName:
+    storageClassName: 
```
```suggestion
    storageClassName:
```

Unintentional change?

jedcunningham commented on chart/templates/rbac/security-context-constraint-rolebinding.yaml:
```diff
@@ -0,0 +1,89 @@
+# under the License.
+
+################################
+## Airflow Pod Reader Role Binding
```
```suggestion
## Airflow SCC Role Binding
```

potiuk commented on chart/values.yaml:
```diff
@@ -1251,7 +1252,7 @@ dags:
     size: 1Gi
     # If using a custom storageClass, pass name here
-    storageClassName:
+    storageClassName: 
```
extra space ;)

nwalens replied:
Yeah, my bad, sorry about that.

potiuk commented on chart/values.schema.json:
```diff
@@ -253,6 +253,11 @@
                     "default": true
+                },
+                "createSCCRoleBinding": {
+                    "description": "Specifies whether SCC RoleBinding resource should be created.",
```
I think it would be great to add a little more explanation of what SCC is. Maybe a small chapter explaining in Production Guide (https://airflow.apache.org/docs/helm-chart/stable/production-guide.html) and link to it from here (and refer to the parameter from there)?  Also a comment that it is related to OpenShift wiould be great.

nwalens replied:
Fair enough, I'm not too good at documentation but I will do my best.

nwalens replied:
Documentation in place!

potiuk commented on chart/templates/rbac/security-context-constraint-rolebinding.yaml:
```diff
@@ -0,0 +1,89 @@
+# under the License.
+
+################################
+## Airflow Pod Reader Role Binding
```
```suggestion
## Airflow Security Context Constraint Role Binding
```

nwalens replied:
Yes, sorry that escaped me.

potiuk commented on chart/templates/rbac/security-context-constraint-rolebinding.yaml:
```diff
@@ -0,0 +1,89 @@
+## Airflow Pod Reader Role Binding
+#################################
+{{- if and .Values.rbac.create .Values.rbac.createSCCRoleBinding }}
+{{- $schedulerLaunchExecutors := list "LocalExecutor" "KubernetesExecutor" "CeleryKubernetesExecutor" }}
```
```suggestion
{{- $schedulerLaunchExecutors := list "LocalExecutor" "CeleryExecutor" "KubernetesExecutor" "CeleryKubernetesExecutor" }}
```

I think ? Or else why "LocalExecutor" is here?

nwalens replied:
Make sense to me as well.

nwalens replied:
Change applied.

potiuk reviewed (commented):
Not sure of all the details but some initial comments

jedcunningham reviewed (commented):
Shouldn't we check that the serviceaccounts are being created?

nwalens committed:
fix typo (extra space)

Signed-off-by: Ney Walens De Mesquita <walens@gmail.com>

nwalens committed:
remove changelog

- The feature will be released in a future release

Signed-off-by: Ney Walens De Mesquita <walens@gmail.com>

nwalens committed:
Check for service accounts for each of the components

- This will ensure that the correct values for each service account are properly checked

Signed-off-by: Ney Walens De Mesquita <walens@gmail.com>

nwalens committed:
Add documentation and improve reference to the new feature.

Signed-off-by: Ney Walens De Mesquita <walens@gmail.com>

jedcunningham commented on chart/templates/rbac/security-context-constraint-rolebinding.yaml:
```diff
@@ -0,0 +1,93 @@
+#################################
+{{- if and .Values.rbac.create .Values.rbac.createSCCRoleBinding }}
+{{- $schedulerLaunchExecutors := list "LocalExecutor" "CeleryExecutor" "KubernetesExecutor" "CeleryKubernetesExecutor" }}
+{{- $workerLaunchExecutors := list "CeleryExecutor" "KubernetesExecutor" "CeleryKubernetesExecutor" }}
```
In this context, I'm not sure we care if the scheduler or workers can launch pods, only if the component exists? Obviously we need a scheduler, so no conditional there. Maybe this is what we want instead for the workers:

```suggestion
{{- $hasWorkers := has .Values.executor (list "CeleryExecutor" "KubernetesExecutor" "CeleryKubernetesExecutor") }}
```

Then the worker serviceaccount section can be done conditionally based on `$hasWorkers`?

nwalens replied:
Yes, I agree, it does look much better like that.
As for schedulers, I think we can remove the test altogether since we always have one in place.

jedcunningham commented on docs/helm-chart/production-guide.rst:
```diff
@@ -200,3 +200,24 @@ By default, the chart will deploy Redis. However, you can use any supported Cele
+When defining a SCC, one can control actions and resources a POD can perform or access during startup and runtime.
+
+The SCCs are split into different levels or categories with the ``restricted`` SCC being the default one assigned to PODs.
+When deploying airflow to OpenShift, one can leverage the SCCs and allow the PODs to start containers utilizing the ``anyuid`` SCC.
```
```suggestion
The SCCs are split into different levels or categories with the ``restricted`` SCC being the default one assigned to Pods.
When deploying Airflow to OpenShift, one can leverage the SCCs and allow the Pods to start containers utilizing the ``anyuid`` SCC.
```

nwalens replied:
Changes in place.

jedcunningham commented on docs/helm-chart/production-guide.rst:
```diff
@@ -200,3 +200,24 @@ By default, the chart will deploy Redis. However, you can use any supported Cele
+-----------------------------
+
+A ``Security Context Constraint`` (SCC) is a OpenShift construct that works as a RBAC rule however it targets PODs instead of users.
+When defining a SCC, one can control actions and resources a POD can perform or access during startup and runtime.
```
```suggestion
A ``Security Context Constraint`` (SCC) is a OpenShift construct that works as a RBAC rule however it targets Pods instead of users.
When defining a SCC, one can control actions and resources a Pod can perform or access during startup and runtime.
```

nwalens replied:
Changes in place.

jedcunningham commented on docs/helm-chart/production-guide.rst:
```diff
@@ -200,3 +200,24 @@ By default, the chart will deploy Redis. However, you can use any supported Cele
+    create: true
+    createSCCRoleBinding: true
+
+In this chart, SCCs are bound to the PODs via RoleBindings meaning that the option ``rbac.create`` must also be set to ``true`` in order to fully enable the SCC usage.
```
```suggestion
In this chart, SCCs are bound to the Pods via RoleBindings meaning that the option ``rbac.create`` must also be set to ``true`` in order to fully enable the SCC usage.
```

nwalens replied:
Changes in place.

jedcunningham reviewed (commented):
@nwalens, I apologize, but looking again I think I was wrong with checking `<component>.serviceAccount.create`. I think we want them to be set regardless and the `<component>.serviceAccountName` templates handles the details. Sorry about that!

nwalens committed:
remove service accunt checks and fix documentation typos.

Signed-off-by: Ney Walens De Mesquita <walens@gmail.com>

nwalens committed:
Merge branch 'apache:main' into main

nwalens committed:
add unit test for SCC creation

Signed-of-by: Ney Walens De Mesquita <walens@gmail.com>

nwalens committed:
fix unit test code formating for static checks

nwalens committed:
Fix test code formating
